### PR TITLE
ENH: Add O(N) random matrix generator

### PR DESCRIPTION
**Pull request by philipdeboer**
Add random O(N) matrices. This is a variation of the SO(N) code, recently merged PR #5622.

### CODE REVIEW

**Comment by philipdeboer** [comment]
How do I update the release notes?

**Comment by ev-br** [comment]
https://github.com/scipy/scipy/blob/master/doc/release/0.18.0-notes.rst

File: scipy/stats/_multivariate.py [hunk 1]
```diff
@@ -2795,3 +2796,109 @@ def __init__(self, dim=None, seed=None):
...
+        Draw random samples from O(N).
+
+    Parameters
+    ----------
+    dim : scalar
+        Dimension of matrices
+
+    Notes
+    ----------
+    This class is closely related to special_ortho_group.
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: scipy.stats.special_ortgho_group wrapped into single backicks would produce a hyperlink in the html docs

File: scipy/stats/_multivariate.py [hunk 2]
```diff
@@ -2795,3 +2796,109 @@ def __init__(self, dim=None, seed=None):
...
+
+    Some care is taken to avoid numerical error, as per the paper by Mezzadri.
+
+    References
+    ----------
+    .. [1] F. Mezzadri, "How to generate random matrices from the classical
+           ompact groups", arXiv:math-ph/0609050v2.
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: Typo: ompact

File: scipy/stats/tests/test_multivariate.py [hunk 3]
```diff
@@ -1091,6 +1091,63 @@ def test_haar(self):
...
+        expected = np.array([[0.993945, -0.045279, 0.100114],
+                             [-0.048216, -0.998469, 0.02711],
+                             [-0.098734, 0.031773, 0.994607]])
+        assert_array_almost_equal(x, expected)
+        assert_almost_equal(np.linalg.det(x), -1)
+
+    def test_invalid_dim(self):
+        assert_raises(ValueError, ortho_group.rvs, None)
+        assert_raises(ValueError, ortho_group.rvs, (2, 2))
+        assert_raises(ValueError, ortho_group.rvs, 1)
```
Comment (ev-br) [inline scipy/stats/tests/test_multivariate.py]: Can you add a test for rvs(2.5) here?

**Comment by ev-br** [comment]
Looks good modulo nitpicks.

**Comment by philipdeboer** [comment]
Updated.

Evgeni, thanks for all your help; it has been a pleasure working on this submission.

Next up are random correlation matrices, but I need to review that code and tidy it up a bit.

**Comment by ev-br** [comment]
CI failures seem real.

**Comment by ev-br** [comment]
One thing I wonder about both this an SO(N) is performance. When I first saw this, I thought I could use it for, e.g. Monte Carlo simulations of a spin model (yeah, I do these things now and then). But then, this code seems heavy on list comprehensions  --- both handling the size argument of rvs, and in the generator itself. I actually wonder if they would not bottleneck the whole simulation. I know I should not FUD before profiling, so let me phrase it like this: have you by any chance done some profiling?
It seems possible that this code can be vectorized if needed.

**Comment by philipdeboer** [comment]
I have not done any profiling. What is a good tool, and what could I benchmark against?

I thought list comprehensions were Pythonic.

For the size arg, I bet we can pre-allocate memory or vectorize this in some way.

The loop in the generator is over different dimensions; eg it is not obvious how to reduce to one the calls to the random number generator. Perhaps pull an array of the correct length, and index into it. I'm open to ideas.

Sent from my BlackBerry device on the Rogers Wireless Network

-----Original Message-----

**Comment by ev-br** [comment]
List comprehensions *are* pythonic, I agree. One possible issue is that pythonic and good performance-wise are not always fully compatible.

Recommended profiling tools for pure python code are `cProfile` from the standard library and `line_profiler` (resp, ipython's magic `%prun` and `%lprun`).

All this is not supposed to be blocking this PR, and can be left for a future PR (or just left as is if it turns that the benefit does not outweigh the increase in code complexity).

**Comment by ev-br** [comment]
Okay, perf improvement can  be done as a follow-up. 
Merging, thanks Philip!

**Comment by Gattocrucco** [comment]
Continuing discussion from #16344, @philipdeboer:

I think that the implementation for O(N) random matrix generation could be rewritten like that of U(N), i.e., using QR, for vastly improved performance, instead of mimicking the SO(N) one. We are in doubt that there were some subtle statistical reasons you had for using that approach. Could you comment on this?

**Comment by philipdeboer** [comment]
I don't recall what the issue may have been. I believe there were some
statistical tests to ensure that the output has the desired properties; if
the more efficient algorithm still passes those tests it should be OK.

**Pull request merged**